### PR TITLE
Patch request gas and bump gas limit up

### PR DESCRIPTION
**Pull request by thebevrishot**
Fixed: #XXXX

## Please ensure the following requirements are met before submitting a pull request:

- [ ] The pull request is targeted against the correct target branch
- [ ] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [ ] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any and all appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [ ] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

**Review by taobun — CHANGES_REQUESTED** [review]
And fix merge conflict and push again.

File: x/oracle/types/msgs.go [hunk 1]
```diff
@@ -74,6 +82,12 @@ func (msg MsgRequestData) ValidateBasic() error {
 	if len(msg.ClientID) > MaxClientIDLength {
 		return WrapMaxError(ErrTooLongClientID, len(msg.ClientID), MaxClientIDLength)
 	}
+	if msg.PrepareGas <= 0 {
+		return sdkerrors.Wrapf(ErrInvalidVersion, "invalid prepare gas: %d", msg.PrepareGas)
```
Comment (Benzbeeb) [inline x/oracle/types/msgs.go:86]: ErrInvalidOwasmGas

File: x/oracle/types/msgs.go [hunk 2]
```diff
@@ -74,6 +82,12 @@ func (msg MsgRequestData) ValidateBasic() error {
...
+	}
+	if msg.ExecuteGas <= 0 {
+		return sdkerrors.Wrapf(ErrInvalidVersion, "invalid execute gas: %d", msg.ExecuteGas)
```
Comment (Benzbeeb) [inline x/oracle/types/msgs.go:89]: ditto

File: x/oracle/types/msgs.go (hunk 1 above)
Comment (Benzbeeb) [reply x/oracle/types/msgs.go:86]: please add test for this error

File: x/oracle/keeper/owasm.go [hunk 3]
```diff
@@ -12,6 +12,13 @@ import (
 	"github.com/bandprotocol/chain/x/oracle/types"
 )
 
+// 1 cosmos gas is equal to 7 owasm gas
+const gasConversionFactor = 7
+
+func convertToOwasmGas(cosmos uint64) uint32 {
+	return uint32(cosmos * gasConversionFactor)
```
Comment (Benzbeeb) [inline x/oracle/keeper/owasm.go:19]: can we use cosmos type `Uint64` for safe multiplication

**Review by Benzbeeb — CHANGES_REQUESTED** [review]

File: hooks/emitter/oracle.go [hunk 4]
```diff
@@ -37,6 +37,7 @@ func (h *Hook) emitOracleModule(ctx sdk.Context) {
 			"tx_hash":          nil,
 			"client_id":        req.ClientID,
 			"resolve_status":   types.RESOLVE_STATUS_OPEN,
+			"execute_gas":      req.ExecuteGas,
```
[the remark this answers is not in the document]
Comment (thebevrishot) [reply hooks/emitter/oracle.go:40]: Added a task in jira

File: x/oracle/keeper/owasm.go (hunk 3 above)
Comment (thebevrishot) [reply x/oracle/keeper/owasm.go:19]: Prepare and execute functions accept uint32 so we have to cast it anyway

File: x/oracle/types/msgs.go (hunk 1 above)
Comment (thebevrishot) [reply x/oracle/types/msgs.go:86]: yep, I'm already append test for this in `TestMsgRequestDataValidation`

**Review by Benzbeeb — APPROVED** [review]

File: x/oracle/client/cli/tx.go [hunk 5]
```diff
@@ -117,6 +131,8 @@ $ %s tx oracle request 1 4 3 --calldata 1234abcdef --client-id cliend-id --from
 
 	cmd.Flags().BytesHexP(flagCalldata, "c", nil, "Calldata used in calling the oracle script")
 	cmd.Flags().StringP(flagClientID, "m", "", "Requester can match up the request with response by clientID")
+	cmd.Flags().Uint64(flagPrepareGas, 0, "Prepare gas used in fee counting for prepare request")
```
Comment (taobun) [inline x/oracle/client/cli/tx.go:134]: Change default value to normal case

File: proto/oracle/v1/tx.proto [hunk 6]
```diff
@@ -67,6 +67,10 @@ message MsgRequestData {
   ];
   // Sender is the sender of this message.
   string sender = 7;
+  // PrepareGas is amount of gas to pay to prepare raw requests
```
Comment (taobun) [inline proto/oracle/v1/tx.proto:70]: Can you move sender to last variable?

File: x/oracle/types/packets_test.go [hunk 7]
```diff
@@ -27,9 +28,12 @@ func TestGetBytesRequestPacket(t *testing.T) {
 		MinCount:       1,
 		FeeLimit:       sdk.NewCoins(sdk.NewCoin("uband", sdk.NewInt(10000))),
 		RequestKey:     "TEST_KEY",
+		PrepareGas:     100,
+		ExecuteGas:     100,
 	}
+	fmt.Println(string(req.GetBytes()))
```
Comment (taobun) [inline x/oracle/types/packets_test.go:34]: Remove

File: x/oracle/types/msgs.go [hunk 8]
```diff
@@ -34,7 +34,13 @@ var (
...
 // NewMsgRequestData creates a new MsgRequestData instance.
 func NewMsgRequestData(
-	oracleScriptID OracleScriptID, calldata []byte, askCount, minCount uint64, clientID string, feeLimit sdk.Coins, sender sdk.AccAddress,
+	oracleScriptID OracleScriptID,
+	calldata []byte,
+	askCount, minCount uint64,
+	clientID string,
+	feeLimit sdk.Coins,
+	sender sdk.AccAddress,
+	prepareGas, executeGas uint64,
```
Comment (taobun) [inline x/oracle/types/msgs.go:43]: Can you swap prepareGas and executeGas before sender?

**Review by taobun — CHANGES_REQUESTED** [review]

File: x/oracle/client/cli/tx.go (hunk 5 above)
Comment (thebevrishot) [reply x/oracle/client/cli/tx.go]: Updated

File: x/oracle/testapp/util.go [hunk 9]
```diff
@@ -2,9 +2,78 @@ package testapp
...
+
+// GasMeterWrapper wrap gas meter for testing purpose
+type GasMeterWrapper struct {
+	sdk.GasMeter
+	GasRecords []GasRecord
+}
+
+func (m *GasMeterWrapper) GasConsumed() sdk.Gas {
+	return m.GasMeter.GasConsumed()
+}
```
Comment (taobun) [inline x/oracle/testapp/util.go:27]: Can override the only function that we want to do?

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**